matgen/generate_matrix_he_and_tz.cc

```diff
@@ -0,0 +1,323 @@
+
+#include "slate/slate.hh"
+#include "slate/generate_matrix.hh"
+#include "random.hh"
+#include "generate_matrix_utils.hh"
+#include "generate_sigma.hh"
+#include "generate_type_rand.hh"
+
+#include <exception>
+#include <string>
+#include <vector>
+#include <limits>
+#include <complex>
+#include <chrono>
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <utility>
+
+namespace slate {
+
+// -----------------------------------------------------------------------------
+/// Generates an m-by-n trapezoid-storage test matrix.
+/// Handles Trapezoid, Triangular, Symmetric, and Hermitian matrices.
+/// @see generate_matrix
+/// @ingroup generate_matrix
+///
+template <typename scalar_t>
+void generate_matrix(
+    MatgenParams& params,
+    slate::BaseTrapezoidMatrix<scalar_t>& A,
+    std::vector< blas::real_type<scalar_t> >& Sigma,
+    slate::Options const& opts)
+{
+    using entry_type = std::function< scalar_t (int64_t, int64_t) >;
+    using real_t = blas::real_type<scalar_t>;
+
+    // Constants
+    const real_t nan = std::numeric_limits<real_t>::quiet_NaN();
+    const real_t d_zero = 0;
+    const real_t d_one  = 1;
+    const scalar_t zero = 0;
+    const scalar_t one  = 1;
+
+    // ----------
+    // set Sigma to unknown (nan)
+    lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                   nan, nan, Sigma.data(), Sigma.size() );
+
+    char msg[ 256 ];
+    TestMatrixType type;
+    TestMatrixDist dist;
+    real_t cond;
+    real_t condD;
+    real_t sigma_max;
+    bool dominant;
+    int64_t zero_col;
+    decode_matrix<scalar_t>(
+        params, A, type, dist, cond, condD, sigma_max, dominant, zero_col );
+
+    int64_t seed = configure_seed(A.mpiComm(), params.seed);
+
+    int64_t n = A.n();
+    int64_t nt = A.nt();
+    int64_t mt = A.mt();
+
+    // ----- generate matrix
+    switch (type) {
+        case TestMatrixType::zeros:
+            set(zero, zero, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_zero, d_zero, Sigma.data(), Sigma.size() );
+            break;
+
+        case TestMatrixType::ones:
+            set(one, one, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_zero, d_one, Sigma.data(), Sigma.size() );
+            if (Sigma.size() >= 1) {
+                Sigma[0] = n;
+            }
+            break;
+
+        case TestMatrixType::identity:
+            set(zero, one, A);
+            lapack::laset( lapack::MatrixType::General, Sigma.size(), 1,
+                d_one, d_one, Sigma.data(), Sigma.size() );
+            break;
+
+        case TestMatrixType::jordan: {
+	    entry_type jordan_entry = [A]( int64_t i, int64_t j  ) {
+                if (A.uplo() == Uplo::Lower) {	
+                    return ( i == j || i + 1 == j ? 1.0 : 0.0 );
+                }
+		else { // upper
+                    return ( i == j || i + 1 == j ? 1.0 : 0.0 );
```

Comment:
Should this be `i - 1 == j`? Or are upper & lower the same?

Reply:
[Jordan matrix](https://en.wikipedia.org/wiki/Jordan_matrix) is lambda on diag, ones on superdiag. We use lambda = 1.
(Jordan matrix)^T is what LAPACK tester uses, with ones on subdiag.
I'm adding both types, `jordan` and `jordanT`.